msprime/ancestry.py

```diff
@@ -721,113 +762,375 @@ def simulate(
+    # make sense. However, having different defaults for ploidy will lead to
+    # confusing problems when we have mixed models, so perhaps we should raise
+    # an error from models that require ploidy=2 and force them to explicitly
+    # set it.
+    # TODO Might also be worth considering setting ploidy=2 by default. This
+    # *is* likely to be the most common application, and those doing haploid
+    # stuff can set ploidy=1, if they like.
+    ploidy = 1 if ploidy is None else ploidy
+    if not core.isinteger(ploidy):
+        raise TypeError("ploidy must be an integer")
+    ploidy = int(ploidy)
+    if ploidy < 1:
+        raise ValueError("ploidy must be >= 1")
+
+    model, model_change_events = _parse_model_arg(model)
+    is_dtwf = isinstance(model, DiscreteTimeWrightFisher)
+
+    # Check the demography. If no demography is specified, we default to a
+    # single-population model with a given population size.
+    if demography is None:
+        if is_dtwf:
+            # A default size of 1 isn't so smart for DTWF and almost certainly
+            # an error.
+            if population_size is None:
                 raise ValueError(
-                    "Must specify a sequence_length for simulations of a "
-                    "genome with recombination or gene conversion without "
-                    "rate map"
+                    "When using the DTWF model, the population size must be set "
+                    "explicity, either using the population_size or demography "
+                    "arguments."
                 )
+        population_size = 1 if population_size is None else float(population_size)
+        demography = demog.Demography.simple_model(population_size)
+    elif isinstance(demography, demog.Demography):
+        if population_size is not None:
+            raise ValueError("Cannot specify demography and population size")
+    else:
+        raise TypeError("demography argument must be an instance of msprime.Demography")
+    # TODO should we validate that the population sizes are positive integers
+    # for DTWF? It's not actually required, but could help catch some errors?
+    demography.validate()
 
-    sim = simulator_factory(
-        sample_size=sample_size,
-        samples=samples,
-        ploidy=ploidy,
-        length=sequence_length,
+    if initial_state is None:
+        if samples is None:
+            raise ValueError("Either samples or initial state must be specified")
+        initial_state = tskit.TableCollection(sequence_length)
+        demography.insert_populations(initial_state)
+        _parse_samples(samples, ploidy, initial_state)
+    else:
+        if samples is not None:
+            raise ValueError("Cannot specify both samples and initial_state")
+        if sequence_length != initial_state.sequence_length:
+            raise ValueError(
+                "The initial_state sequence length must agree with other parameters"
```

Comment:
It might help the user to see a list of other parameters that this must agree with. Although I suppose such a list might get longer in the future (e.g if a mutation map is added), so this risks becoming outdated and thus confusing if the problem parameter isn't in the list.